### ISSUE TITLE
Language feature for parametric formulations security

### ISSUE THREAD

**Issue by modelica-trac-importer** [tbeu/ModelicaSpecification#4]
**Modified by dietmarw on 4 Feb 2008 05:29 UTC**
Some time ago, we did some playing around with VHDL-AMS.  I thought a clutch model would be a good way to see how expressive the language was and how robust the underlying tools were.  So we created a very simple minded parametric clutch model that was basically expressed as:

```text/x-modelica
if s>tau_max then
  tau := tau_max;
  w_rel = tau_max-s;
elseif s<-tau_max then
  tau := -tau_max;
  w_rel = s-tau_max;
else
  tau := s;
  w_rel = 0;
end if;
```

I did this thinking that such a simple minded approach (compared to the friction models in the Modelica Standard Library) couldn't possibly work but at least it would establish a starting point.  To my surprise, this did work.  Not only that, we validated the numerical results against Dymola.  Even more surprisingly, the VHDL-AMS tool was only using Euler integration and it didn't show any numerical instabilities.

In light of this, I think we need to patch this issue or else we risk looking silly when people compare VHDL-AMS and Modelica models of such simple behavior.

I appreciate that different solution techniques ( e.g. index reduction) may have difficulty with such formulations although clearly they are possible.  As such, my proposal would be to introduce a high-level way to express such parametric relationships.  These come up all the time.  The need to express two variables parametrically in terms of a third can be found in a standard clutch, a one way clutch, a diode, a cavitating hydraulic piston, contact problems, etc.

As such, I think the way to address this is to introduce a new built-in function that will greatly simplify these types of expressions while, at the same time, allowing tools to recognize this relationship explicitly and optimize as a result ( e.g. perform appropriate scaling).  In my experience, these kinds of relationships, in the context of index reduction, lead to the need to manually formulate a state transition diagram and in some cases even that doesn't lead to a proper solution.  That is why I think we need this to be "built-in" ( i.e. it may require/benefit the tools to do some special handling).

The interface for the function looks like this:

```text/x-modelica
function parametric
  input Real s "Independent variable";
  input Real points[:,2] = [0,0] "Points defining the curve";
  input Integer origin = 1 "Index of the point to consider the origin (i.e. where s=0)";
  input Real leftSlope[2] "Slope (w.r.t s) before the first point";
  input Real rightSlope[2] "Slope ( w.r.t s) after the last point";
  output Real y "One dependent variable";
  output Real x "The other dependent variable";
...
end parametric;
```

Based on my experience, I don't think that making this a block or providing derivative annotations solves the fundamental issues that require it to be a built-in but I'm open to other approaches.

Here are some sample models that I've created to demonstrate the function.  I've actually implemented the 'parametric' function in ordinary Modelica.  The problem is that even though the function provides values as expected, Dymola can't solve these models because of index reduction issues.

```text/x-modelica
model ParametricDiode
  extends Modelica.Electrical.Analog.Interfaces.OnePort;
protected
  Real s;
equation
  (i, v) = parametric(s=s,leftSlope={0,1},
                          rightSlope={1,0});
end ParametricDiode;

model ParametricClutch
  parameter Real mu;
  parameter Modelica.SIunits.Area A;
  extends Modelica.Mechanics.Rotational.Interfaces.Compliant;
  Modelica.Blocks.Interfaces.RealInput f_normalized;
protected
  Real s "Parametric variable";
  Modelica.SIunits.AngularVelocity w_rel = der(phi_rel);
  Modelica.SIunits.Torque tau_max = max(0,f_normalized)*mu*A;
equation
  (tau, w_rel) = parametric(s=s,points=[-tau_max,0;0,0;tau_max,0],
                            origin=2,
                            leftSlope={0,1},
                            rightSlope={0,1});
end ParametricClutch;

model ParametricOneWayClutch
  parameter Real mu;
  parameter Modelica.SIunits.Area A;
  extends Modelica.Mechanics.Rotational.Interfaces.Compliant;
  Modelica.Blocks.Interfaces.RealInput f_normalized;
protected
  Real s "Parametric variable";
  Modelica.SIunits.AngularVelocity w_rel = der(phi_rel);
  Modelica.SIunits.Torque tau_max = max(0,f_normalized)*mu*A;
equation
  (tau, w_rel) = parametric(s=s,points=[tau_max,0],
                            leftSlope={1,0},
                            rightSlope={0,1});
end ParametricOneWayClutch;
```

I've kept these models relatively simple just so they are clear.  While the standard library models are slightly more complicated (e.g . the diode includes Goff and Ron parameters), these can be included using the 'parametric' function as I've defined it.

It is perhaps worth noting that such a function could be used in the context of contact problems as well.  The only wrinkle there is that they introduce impulses.  The reason I think this is interesting is that it is yet another case where the tool may have to do something intelligent to solve the problem and why it should be a built-in ( e.g. so that impulses that result from the formulation can be handled properly).

The bottom line is that modeling friction is currently too hard in Modelica and we need to give users a better way.

P.S. - I'm attaching the code I was working with while thinking about this.  There are several example models in there (not all run for the reasons I've described).

----

**Reported by mtiller on 31 Jan 2008 20:01 UTC**
Some time ago, we did some playing around with VHDL-AMS.  I thought a clutch model would be a good way to see how expressive the language was and how robust the underlying tools were.  So we created a very simple minded parametric clutch model that was basically expressed as:

```
if s>tau_max then
  tau := tau_max;
  w_rel = tau_max-s;
elseif s<-tau_max then
  tau := -tau_max;
  w_rel = s-tau_max;
else
  tau := s;
  w_rel = 0;
end if;
```

I did this thinking that such a simple minded approach (compared to the friction models in the Modelica Standard Library) couldn't possibly work but at least it would establish a starting point.  To my surprise, this did work.  Not only that, we validated the numerical results against Dymola.  Even more surprisingly, the VHDL-AMS tool was only using Euler integration and it didn't show any numerical instabilities.

In light of this, I think we need to patch this issue or else we risk looking silly when people compare VHDL-AMS and Modelica models of such simple behavior.

I appreciate that different solution techniques ( e.g. index reduction) may have difficulty with such formulations although clearly they are possible.  As such, my proposal would be to introduce a high-level way to express such parametric relationships.  These come up all the time.  The need to express two variables parametrically in terms of a third can be found in a standard clutch, a one way clutch, a diode, a cavitating hydraulic piston, contact problems, etc.

As such, I think the way to address this is to introduce a new built-in function that will greatly simplify these types of expressions while, at the same time, allowing tools to recognize this relationship explicitly and optimize as a result ( e.g. perform appropriate scaling).  In my experience, these kinds of relationships, in the context of index reduction, lead to the need to manually formulate a state transition diagram and in some cases even that doesn't lead to a proper solution.  That is why I think we need this to be "built-in" ( i.e. it may require/benefit the tools to do some special handling).

The interface for the function looks like this:

```
function parametric
  input Real s "Independent variable";
  input Real points[:,2] = [0,0] "Points defining the curve";
  input Integer origin = 1 "Index of the point to consider the origin (i.e. where s=0)";
  input Real leftSlope[2] "Slope (w.r.t s) before the first point";
  input Real rightSlope[2] "Slope ( w.r.t s) after the last point";
  output Real y "One dependent variable";
  output Real x "The other dependent variable";
...
end parametric;
```

Based on my experience, I don't think that making this a block or providing derivative annotations solves the fundamental issues that require it to be a built-in but I'm open to other approaches.

Here are some sample models that I've created to demonstrate the function.  I've actually implemented the 'parametric' function in ordinary Modelica.  The problem is that even though the function provides values as expected, Dymola can't solve these models because of index reduction issues.

```
model ParametricDiode
  extends Modelica.Electrical.Analog.Interfaces.OnePort;
protected
  Real s;
equation
  (i, v) = parametric(s=s,leftSlope={0,1},
                          rightSlope={1,0});
end ParametricDiode;

model ParametricClutch
  parameter Real mu;
  parameter Modelica.SIunits.Area A;
  extends Modelica.Mechanics.Rotational.Interfaces.Compliant;
  Modelica.Blocks.Interfaces.RealInput f_normalized;
protected
  Real s "Parametric variable";
  Modelica.SIunits.AngularVelocity w_rel = der(phi_rel);
  Modelica.SIunits.Torque tau_max = max(0,f_normalized)*mu*A;
equation
  (tau, w_rel) = parametric(s=s,points=[-tau_max,0;0,0;tau_max,0],
                            origin=2,
                            leftSlope={0,1},
                            rightSlope={0,1});
end ParametricClutch;

model ParametricOneWayClutch
  parameter Real mu;
  parameter Modelica.SIunits.Area A;
  extends Modelica.Mechanics.Rotational.Interfaces.Compliant;
  Modelica.Blocks.Interfaces.RealInput f_normalized;
protected
  Real s "Parametric variable";
  Modelica.SIunits.AngularVelocity w_rel = der(phi_rel);
  Modelica.SIunits.Torque tau_max = max(0,f_normalized)*mu*A;
equation
  (tau, w_rel) = parametric(s=s,points=[tau_max,0],
                            leftSlope={1,0},
                            rightSlope={0,1});
end ParametricOneWayClutch;
```

I've kept these models relatively simple just so they are clear.  While the standard library models are slightly more complicated (e.g . the diode includes Goff and Ron parameters), these can be included using the 'parametric' function as I've defined it.

It is perhaps worth noting that such a function could be used in the context of contact problems as well.  The only wrinkle there is that they introduce impulses.  The reason I think this is interesting is that it is yet another case where the tool may have to do something intelligent to solve the problem and why it should be a built-in ( e.g. so that impulses that result from the formulation can be handled properly).

The bottom line is that modeling friction is currently too hard in Modelica and we need to give users a better way.

P.S. - I'm attaching the code I was working with while thinking about this.  There are several example models in there (not all run for the reasons I've described).

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/43

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 1 Feb 2008 07:31 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 4 Feb 2008 05:29 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 4 Feb 2008 07:17 UTC**
**on 01-02-2008 08:53, Martin Otter wrote:**
Dear Michael,
CC: Modelicans

your comment with regards to VHDL-AMS and friction modeling below is questionable. Find attached  [attachment:ticket:43:HybridDAESNeedSymbolicPreProcessing.pdf a section of my book] on which I am currently writing where a much simpler case is sketched that shows that a pure numerical approach cannot work satisfactorily for higher index hybrid DAEs. Only in special cases it might work and you have obviously picked such an example.

I agree with you that friction modeling should be simplified in Modelica. However, it is not a simple topic.

Best regards,

Martin

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 4 Feb 2008 07:23 UTC**
**On 01-02-2008 15:54, Martin Otter wrote:**

I recognized that my ~~email~~ post above<sub>(changed by dietmarw)</sub> could be misleading:

The fundamental problem is that it is generally not possible to integrate numerically a singular DAE system (= index > 1) and handle events. The reason is that re-initialization at an event instant requires to evaluate the DAE equation system, but this equation system is singular, i.e., has an infinite number of solutions.

One solution is to not have events, i.e., replace discontinuities by steep characteristics, but this has other drawbacks.

Simulating a singular DAE system under the presence of events (that may potentially change variables discontinuously, such as forces and torques) requires at the minimum that for re-initialization the "hidden" constraints are utilized. These "hidden" constraints can be computed for certain classes of DAE systems (like multi-body systems). For the general case, where not more information is known as just the equations, the only known method to determine these hidden constraints is the Pantelides algorithm. According to my knowledge, no VHDL-AMS simulator uses the Pantelides algorithm and therefore they are not able to solve such systems (higher index and events).

To summarize:
Every simulation environment that would like to simulate all the systems from the Modelica Standard Library in a reliable way, must support the Pantelides algorithm or a variant of it, since it is easily possible with MSL to build higher index DAEs with events. Solving such systems by pure numerical methods, without symbolic pre-processing, only works if no events are present (that have an impact on the singular system) and if the index of the model is not too high.

Best regards,

Martin

at the same time re-initialize at an event instant

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 6 Mar 2008 17:26 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by otter on 16 Jun 2009 16:34 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by HansOlsson on 26 Jan 2010 16:37 UTC**
Needs more information.

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 23 Jan 2013 09:04 UTC**

**Comment by modelica-trac-importer** [comment]
**Changelog removed by dietmarw on 23 Jan 2013 09:04 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 23 Jun 2016 12:59 UTC**
Housekeeping to make it easier to find the active tickets for the specification.

These tickets are all expired, i.e. old with no action - and not clear that any action is needed. In most cases they are enhancements/discussions without any clear result.
